src/opts.c

```diff
@@ -883,17 +883,190 @@ natsOptions_SetMaxPendingMsgs(natsOptions *opts, int maxPending)
+    for (; this != NULL; this = this->next)
+    {
+        tail = NATS_CALLOC(1, sizeof(nats_CallbackList));
+        if (tail == NULL)
+        {
+            natsOptions_freeCallbackList(head);
+            return nats_setDefaultError(NATS_NO_MEMORY);
+        }
+        *tail = *this;
+        tail->next = NULL;
+
+        if (prev == NULL)
+        {
+            head = tail;
+        }
+        else
+        {
+            prev->next = tail;
+        }
+        prev = tail;
+    }
+
+    *clone = head;
+    return NATS_OK;
+}
+
+natsStatus
+natsOptions_addConnectionClosedCallback(natsOptions *opts, natsConnectionHandler f, void *closure)
+{
+    natsStatus          s           = NATS_OK;
+    nats_CallbackList   *cb         = NULL;
+    nats_CallbackList   *replaced   = NULL;
+
+    cb = NATS_CALLOC(1, sizeof(nats_CallbackList));
+    if (cb == NULL)
+        return nats_setDefaultError(NATS_NO_MEMORY);
+    cb->type = CALLBACK_TYPE_CONN;
+    cb->f.conn = f;
+    cb->closure = closure;
+
+    LOCK_AND_CHECK_OPTIONS(opts, 0);
+
+    // unlink if already in the list.
+    natsOptions_unlinkCallback(&replaced, &opts->closedCb, (void(*)(void))f, closure);
+
+    // add at the head.
+    cb->next = opts->closedCb;
+    opts->closedCb = cb;
+
+    UNLOCK_OPTS(opts);
+
+    NATS_FREE(replaced);
+
+    return s;
+}
+
+natsStatus
+natsOptions_addErrorCallback(natsOptions *opts, natsErrHandler f, void *closure)
+{
+    natsStatus          s           = NATS_OK;
+    nats_CallbackList   *cb         = NULL;
+    nats_CallbackList   *replaced   = NULL;
+
+    cb = NATS_CALLOC(1, sizeof(nats_CallbackList));
+    if (cb == NULL)
+        return nats_setDefaultError(NATS_NO_MEMORY);
+    cb->type = CALLBACK_TYPE_ERROR;
+    cb->f.err = f;
+    cb->closure = closure;
+
+    LOCK_AND_CHECK_OPTIONS(opts, 0);
+
+    // unlink if already in the list.
+    natsOptions_unlinkCallback(&replaced, &opts->asyncErrCb, (void(*)(void))f, closure);
+
+    // add at the head.
+    cb->next = opts->asyncErrCb;
+    opts->asyncErrCb = cb;
+
+    UNLOCK_OPTS(opts);
+
+    return s;
+}
+
+void natsOptions_removeConnectionClosedCallback(natsOptions *opts, natsConnectionHandler f, void *closure)
+{
+    nats_CallbackList *removed = NULL;
+
+    natsMutex_Lock(opts->mu);
+
+    natsOptions_unlinkCallback(&removed, &opts->closedCb, (void(*)(void))f, closure);
+
+    UNLOCK_OPTS(opts);
+
+    NATS_FREE(removed);
+}
+
+void natsOptions_removeErrorCallback(natsOptions *opts, natsErrHandler f, void *closure)
+{
+    nats_CallbackList *removed = NULL;
+
+    natsMutex_Lock(opts->mu);
+
+    natsOptions_unlinkCallback(&removed, &opts->asyncErrCb, (void(*)(void))f, closure);
+
+    UNLOCK_OPTS(opts);
+
+    NATS_FREE(removed);
+}
+
 natsStatus
 natsOptions_SetErrorHandler(natsOptions *opts, natsErrHandler errHandler,
                             void *closure)
 {
-    LOCK_AND_CHECK_OPTIONS(opts, 0);
+    if (errHandler == NULL)
+    {
+        errHandler = natsConn_defaultErrHandler;
```

Comment:
But in this case, should you not keep track if there was an "old" one and if so free it like you do for `SetClosedCB`? In other words, for error CB we always set it to something, but that does not mean that we may be replacing something that would then need to be freed?

Reply:
Actually, no. Here we only allow overriding a "single" callback, and hijack the struct that is already in place. `SetClosedCB` allows to set to `NULL` in which case there is something to free.

Sorry I overreacted and pushed a commit to "fix" this, now